### PR TITLE
#361 Support pkg-config for cflags only 

### PR DESCRIPTION
**Pull request by BNMetrics**
- Added pkg_config_cflags option in cc related build rules in  `/src/parse/rules/cc_rules.build_defs` and `/src/parse/rules/c_rules.build_defs`

### CODE REVIEW

File: src/parse/rules/cc_rules.build_defs [hunk 1]
```diff
@@ -755,18 +771,28 @@ def _default_cflags(c, dbg):
...
 
-def _build_flags(compiler_flags:list, pkg_config_libs:list, defines=None, c=False, dbg=False):
+def _build_flags(compiler_flags:list, pkg_config_libs:list, pkg_config_cflags:list, defines=None, c=False, dbg=False):
     """Builds flags that we'll pass to the compiler invocation."""
     compiler_flags = [_default_cflags(c, dbg), '-fPIC'] + compiler_flags  # N.B. order is important!
     if defines:
         compiler_flags += ['-D' + define for define in defines]
-    pkg_config_cmd = ' '.join([f'`pkg-config --cflags {x}`' for x in pkg_config_libs])
+
+    pkg_config_list = pkg_config_libs or pkg_config_cflags or []
```
Comment (peterebden) [inline src/parse/rules/cc_rules.build_defs]: I'm not sure `or` is correct here. Shouldn't it be `pkg_config_libs + pkg_config_cflags`?

File: src/parse/rules/cc_rules.build_defs [hunk 2]
```diff
@@ -755,18 +771,28 @@ def _default_cflags(c, dbg):
...
+    pkg_config_cmd = ' '.join([f'`pkg-config --cflags {x}`' for x in pkg_config_list])
+
     return ' '.join(compiler_flags) + ' ' + pkg_config_cmd
 
 
-def _binary_build_flags(linker_flags:list, pkg_config_libs:list, shared=False, alwayslink='', c=False, dbg=False):
+def _binary_build_flags(linker_flags:list, pkg_config_libs:list, pkg_config_cflags:list, shared=False, alwayslink='', c=False, dbg=False):
```
Comment (peterebden) [inline src/parse/rules/cc_rules.build_defs]: I don't think it makes sense to pass pkg_config_cflags to this function. They're not going to have any impact on the linker.

File: src/parse/rules/cc_rules.build_defs [hunk 3]
```diff
@@ -822,34 +850,43 @@ def _binary_cmds(c, linker_flags, pkg_config_libs, extra_flags='', shared=False,
...
     def apply_transitive_labels(name):
         labels = get_labels(name, 'cc:')
         flags = ['-isystem %s' % l[4:] for l in labels if l.startswith('inc:')]
         flags.extend(['-D' + l[4:] for l in labels if l.startswith('def:')])
-        pkg_config_libs.extend([l[3:] for l in labels if l.startswith('pc:') and l[3:] not in pkg_config_libs])
+
+        if pkg_config_libs:
+            pkg_config_libs.extend([l[3:] for l in labels if l.startswith('pc:') and l[3:] not in pkg_config_libs])
+        elif pkg_config_cflags:
+            pkg_config_libs.extend([l[3:] for l in labels if l.startswith('pcc:') and l[4:] not in pkg_config_cflags])
```
Comment (peterebden) [inline src/parse/rules/cc_rules.build_defs]: should probably be `l[4:]` in both places

File: src/parse/rules/cc_rules.build_defs [hunk 4]
```diff
@@ -822,34 +850,43 @@ def _binary_cmds(c, linker_flags, pkg_config_libs, extra_flags='', shared=False,
...
-def _library_transitive_labels(c, compiler_flags, pkg_config_libs, archive=True):
+def _library_transitive_labels(c, compiler_flags, pkg_config_libs, pkg_config_cflags, archive=True):
     """Applies commands from transitive labels to a cc_library rule."""
     def apply_transitive_labels(name):
         labels = get_labels(name, 'cc:')
         flags = ['-isystem %s' % l[4:] for l in labels if l.startswith('inc:')]
         flags.extend(['-D' + l[4:] for l in labels if l.startswith('def:')])
-        pkg_config_libs.extend([l[3:] for l in labels if l.startswith('pc:') and l[3:] not in pkg_config_libs])
+
+        if pkg_config_libs:
```
Comment (peterebden) [inline src/parse/rules/cc_rules.build_defs]: I don't think these should be conditional, presumably it should extend `pkg_config_cflags` in the same way as the logic worked before

File: src/parse/rules/cc_rules.build_defs [hunk 5]
```diff
@@ -822,34 +850,43 @@ def _binary_cmds(c, linker_flags, pkg_config_libs, extra_flags='', shared=False,
...
         if flags:  # Don't update if there aren't any relevant labels
-            cmds, _ = _library_cmds(c, compiler_flags, pkg_config_libs, ' '.join(flags), archive=archive)
+            cmds, _ = _library_cmds(c, compiler_flags, pkg_config_libs, pkg_config_cflags, ' '.join(flags), archive=archive)
             for k, v in cmds.items():
                 set_command(name, k, v)
     return apply_transitive_labels
 
 
-def _binary_transitive_labels(c, linker_flags, pkg_config_libs, shared=False):
+def _binary_transitive_labels(c, linker_flags, pkg_config_libs=[], pkg_config_cflags=[], shared=False):
```
Comment (peterebden) [inline src/parse/rules/cc_rules.build_defs]: same here, this function doesn't need `pkg_config_cflags`

File: src/parse/rules/cc_rules.build_defs [hunk 6]
```diff
@@ -822,34 +850,43 @@ def _binary_cmds(c, linker_flags, pkg_config_libs, extra_flags='', shared=False,
...
     """Applies commands from transitive labels to a cc_binary, cc_test or cc_shared_object rule."""
     def apply_transitive_labels(name):
         labels = get_labels(name, 'cc:')
         linker_prefix = '' if CONFIG.LINK_WITH_LD_TOOL else '-Wl,'
         flags = [linker_prefix + l[3:] for l in labels if l.startswith('ld:')]
-        flags.extend(['`pkg-config --libs %s`' % l[3:] for l in labels if l.startswith('pc:')])
+        if pkg_config_cflags:
+            flags.extend(['`pkg-config --cflags %s`' % l[4:] for l in labels if l.startswith('pcc:')])
+        elif pkg_config_libs:
+            flags.extend(['`pkg-config --libs %s`' % l[3:] for l in labels if l.startswith('pc:')])
```
Comment (peterebden) [inline src/parse/rules/cc_rules.build_defs]: and this line shouldn't be conditional

File: test/cc_rules/clang/BUILD [hunk 7]
```diff
@@ -52,7 +52,7 @@ cc_library(
 cc_shared_object(
     name = "so_test",
     srcs = ["so_test.cc"],
-    pkg_config_libs = ["python3"],
+    pkg_config_cflags = ["python3"],
```
Comment (peterebden) [inline test/cc_rules/clang/BUILD]: You might need to add something like `ldflags = "-undefined dynamic_lookup"` on OSX; according to https://blog.tim-smith.us/2015/09/python-extension-modules-os-x/ that seems to be needed (the OSX linker is a constant source of little differences like this)

Comment (BNMetrics) [reply test/cc_rules/clang/BUILD]: I tried to append this linker flag on the actually build_rule, however, it doesn't seem to have worked... :(

File: src/parse/rules/c_rules.build_defs [hunk 8]
```diff
@@ -27,8 +27,10 @@ def c_library(name:str, srcs:list=[], hdrs:list=[], private_hdrs:list=[], deps:l
       compiler_flags (list): Flags to pass to the compiler.
       linker_flags (list): Flags to pass to the linker; these will not be used here but will be
                            picked up by a c_binary or c_test rule.
-      pkg_config_libs (list): Libraries to declare a dependency on using pkg-config. Again, the ldflags
-                              will be picked up by c_binary or c_test rules.
+      pkg_config_libs (list): Libraries to declare a dependency on using `pkg-config --libs`. Again, the ldflags
```
Comment (peterebden) [inline src/parse/rules/c_rules.build_defs]: This isn't quite correct; it does also imply --cflags sometimes too.

Suggest leaving this one as it was before.

File: src/parse/rules/c_rules.build_defs [hunk 9]
```diff
@@ -27,8 +27,10 @@ def c_library(name:str, srcs:list=[], hdrs:list=[], private_hdrs:list=[], deps:l
...
+                              will be picked up by cc_binary or cc_test rules.
+      pkg_config_cflags (list): Libraries to declare a dependency on using `pkg-config --cflags`. Again, the ldflags
```
Comment (peterebden) [inline src/parse/rules/c_rules.build_defs]: The second sentence is misleading (there are no ldflags in this case). Maybe just leave it out.

File: test/cc_rules/clang/BUILD (hunk 7 above)
Comment (peterebden) [reply test/cc_rules/clang/BUILD]: Can you change this one to be like the other one too please? (i.e. do the CONFIG.OS check)

**Pull request merged**